rapthor/lib/cluster.py

```diff
@@ -3,118 +3,178 @@
-        Bandwidth in Hz per frequency sample
-    solint_slow_freqstep : int
-        Number of frequency samples in slow-gain solve
-    solint_slow_timestep : int
-        Number of time samples in slow-gain solve
+    numsamples : int
+        Total number of frequency samples in the observation
+    numobs : int
+        Total number of observations
+    target_freqstep : int
+        Target number of frequency samples for the slow-gain solve
+    target_timestep : int
+        Target number of time samples for the slow-gain solve
     antenna : str
         Antenna type: "HBA" or "LBA"
     ndir : int
         Number of directions/patches in the calibration
+
+    Returns
+    -------
+    samples_per_chunk : int
+        Size of chunk in number of frequency channels
+    solint : int
+        Solution interval in number of frequency channels that will ensure the solve
+        fits in the available memory
     """
-    # Try to make at least as many time chunks as there are nodes
-    mem_gb = get_total_memory()
-    if antenna == 'HBA':
-        # Memory usage in GB/chan/timeslot/dir of a typical HBA observation with
-        # only Dutch stations
-        mem_usage_gb = 0.83e-3
-    elif antenna == 'LBA':
-        # Memory usage in GB/chan/timeslot/dir of a typical LBA observation
-        mem_usage_gb = 2.5e-4
-    gb_per_solint = mem_usage_gb * solint_slow_freqstep * solint_slow_timestep * ndir
-
-    # Determine if we need to reduce solint_slow_timestep to fit in memory. We adjust
-    # the time step rather than the frequency step, as it is less critical and usually
-    # has a finer sampling
-    if mem_gb / gb_per_solint < 1.0:
-        old_solint_slow_timestep = solint_slow_timestep
-        solint_slow_timestep *= mem_gb / gb_per_solint
-        solint_slow_timestep = max(1, int(round(solint_slow_timestep)))
+    def get_gb_per_solution(target_freqstep, target_timestep, ndir):
+        # Estimate the memory usage in GB per solution
+        #
+        # Note: the numbers below were determined empirically from typical (Dutch-only)
+        # datasets
+        if antenna == 'HBA':
+            # Best-fit coefficients for 25 time slots
+            coef = [7.33333333e-01, 8.33333333e-02, 7.66666667e-03, 5.50000000e-03,
+                    -3.20833714e-17, -4.39443416e-18]
+        elif antenna == 'LBA':
+            # TODO: determine best-fit coefficients
+            # (for now set to HBA values)
+            coef = [7.33333333e-01, 8.33333333e-02, 7.66666667e-03, 5.50000000e-03,
+                    -3.20833714e-17, -4.39443416e-18]
+        gb_per_25_timeslots = (coef[4]*ndir**2 + coef[5]*target_freqstep**2 +
+                               coef[3]*ndir*target_freqstep + coef[1]*ndir +
+                               coef[2]*target_freqstep + coef[0])
+
+        # Return total usage, scaled to the requested time interval, plus a 10% safety
+        # buffer
+        return gb_per_25_timeslots * target_timestep / 25 * 1.1
```

Comment:
This might need some explanation. I don't know what these coefficients are used for, nor do I understand the formula you're using to calculate `gb_per_25_timeslots`.

Reply:
I added some more info regarding the fitting that was done.